src/validator/task.rs

```diff
@@ -0,0 +1,942 @@
+fn _disconnect_block(
+    _rwtxn: &mut RwTxn,
+    _dbs: &Dbs,
+    event_tx: &Sender<Event>,
+    block_hash: Hash256,
+) -> Result<(), DisconnectBlockError> {
+    // FIXME: implement
+    todo!();
+    let event = Event::DisconnectBlock { block_hash };
+    let _send_err: Result<Option<_>, TrySendError<_>> = event_tx.try_broadcast(event);
+    Ok(())
+}
+
+fn _is_transaction_valid(
+    _rotxn: &mut RoTxn,
+    _dbs: &Dbs,
+    _transaction: &Transaction,
+) -> Result<(), TxValidationError> {
+    todo!();
+}
+
+fn initial_sync(
+    dbs: &Dbs,
+    event_tx: &Sender<Event>,
+    main_client: &RpcClient,
+) -> Result<(), InitialSyncError> {
+    let mut rwtxn = dbs.write_txn()?;
+    let mut height = dbs
+        .current_block_height
+        .try_get(&rwtxn, &UnitKey)?
+        .unwrap_or(0);
+    let main_block_height: u32 = main_client
+        .send_request("getblockcount", &[])
+        .map_err(|err| InitialSyncError::Rpc {
+            method: "getblockcount".to_owned(),
+            source: Box::new(err),
+        })?
+        .ok_or(InitialSyncError::GetBlockCount)?;
+    tracing::debug!("mainchain block height: {main_block_height}");
+    if height < main_block_height {
+        tracing::debug!("syncing to block height {main_block_height}");
+    }
+    while height < main_block_height {
+        let block_hash: String = main_client
+            .send_request("getblockhash", &[json!(height)])
+            .map_err(|err| InitialSyncError::Rpc {
+                method: "getblockhash".to_owned(),
+                source: Box::new(err),
+            })?
+            .ok_or(InitialSyncError::GetBlockHash { height })?;
+        let block: String = main_client
+            .send_request("getblock", &[json!(block_hash), json!(0)])
+            .map_err(|err| InitialSyncError::Rpc {
+                method: "getblock".to_owned(),
+                source: Box::new(err),
+            })?
+            .ok_or_else(|| InitialSyncError::GetBlock {
+                block_hash: block_hash.clone(),
+            })?;
+        let block = bitcoin::consensus::encode::deserialize_hex(&block).unwrap();
+        let () = connect_block(&mut rwtxn, dbs, event_tx, &block, height)?;
+        {
+            /*
+            main_client
+                .send_request("invalidateblock", &[json!(block_hash)])
+                .into_diagnostic()?
+                .ok_or(miette!("failed to invalidate block"))?;
+            */
+        }
+        height += 1;
+        let block_hash = {
+            match hex::decode(&block_hash) {
+                Ok(block_hash) => block_hash,
+                Err(err) => {
+                    return Err(InitialSyncError::DecodeBlockHashHex {
+                        block_hash_hex: block_hash,
+                        source: err,
+                    })
+                }
+            }
+            .try_into()
+            .map(BlockHash::from_byte_array)
+            .unwrap()
+        };
```

Comment:
This is incorrect. Block hashes must be deserialised in reverse! There's an implementation of this on the BlockHash type. 

```rust
let block_hash = {
            match BlockHash::from_str(&block_hash) {
                Ok(block_hash) => block_hash,
                Err(err) => {
                    return Err(InitialSyncError::DecodeBlockHashHex {
                        block_hash_hex: block_hash,
                        source: err,
                    })
                }
            }
        };
```

Reply:
Which also means that `source` in `InitialSyncError.DecodeBlockHashHex` must be of type `HexToArrayError`

Reply:
Nice catch, will fix this

Reply:
Fixed